src/install/dracut-install.c

```diff
@@ -426,14 +427,82 @@ static int cp(const char *src, const char *dst)
         return ret;
 }
 
+static int decompress_cp(const char *src, const char *dst)
+{
+        char decompressor[5];
+        char *dest_base = strdup(dst);
+        if (endswith(src, ".xz")) {
+                strcpy(decompressor, "xz");
+                dest_base[strlen(dest_base) - 3] = 0;
+
+        } else if (endswith(src, ".gz")) {
+                strcpy(decompressor, "gz");
+                dest_base[strlen(dest_base) - 3] = 0;
+
+        } else if (endswith(src, ".zst")) {
+                strcpy(decompressor, "zstd");
+                dest_base[strlen(dest_base) - 4] = 0;
+        } else {
+                if (endswith(src, ".ko")) {
+                        log_debug("not compressed: %s", src);
+                } else {
+                        log_warning("unknown compression: %s", src);
+                }
+                // Not compressed or unknown compressor, fallback to copying
+                goto fallback;
+        }
+
+        int pid;
+        int ret = 0;
+        struct stat sb;
+
+        if (lstat(src, &sb) != 0)
+                goto fallback;
+
+        int dest_desc =
+                open(dest_base, O_WRONLY | O_CREAT | O_CLOEXEC | O_TRUNC,
+                     (sb.st_mode) & (S_ISUID | S_ISGID | S_ISVTX | S_IRWXU | S_IRWXG | S_IRWXO));
+        if (dest_desc < 0)
+                goto fallback;
+
+        pid = fork();
+        if (pid == 0) {
+                dup2(dest_desc, 1);
+                execlp(decompressor, decompressor, "--keep", "--decompress", src, "--stdout", "--force", NULL);
```

Comment:
Not a big matter, but `gzip` an `xz` man pages put the source file in the end of the command line.